### PR TITLE
JDK-8301496: Replace NULL with nullptr in cpu/riscv

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi, this PR changes all occurrences of NULL to nullptr for the subdirectory cpu/riscv. Unfortunately the script that does the change isn't perfect, and so we
need to comb through these manually to make sure nothing has gone wrong. I also review these changes but things slip past my eyes sometimes.

Here are some typical things to look out for:

1. No changes but copyright header changed (probably because I reverted some changes but forgot the copyright).
2. Macros having their NULL changed to nullptr, these are added to the script when I find them. They should be NULL.
3. nullptr in comments and logs. We try to use lower case "null" in these cases as it reads better. An exception is made when code expressions are in a comment.

An example of this:

```c++
// This function returns null
void* ret_null();
// This function returns true if *x == nullptr
bool is_nullptr(void** x);
```

Note how `nullptr` participates in a code expression here, we really are talking about the specific value `nullptr`.

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301496](https://bugs.openjdk.org/browse/JDK-8301496): Replace NULL with nullptr in cpu/riscv

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [cf24946e](https://git.openjdk.org/jdk/pull/12324/files/cf24946e788214e05e6b48c01dc6419e3bbcce68)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12324/head:pull/12324` \
`$ git checkout pull/12324`

Update a local copy of the PR: \
`$ git checkout pull/12324` \
`$ git pull https://git.openjdk.org/jdk.git pull/12324/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12324`

View PR using the GUI difftool: \
`$ git pr show -t 12324`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12324.diff">https://git.openjdk.org/jdk/pull/12324.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12324#issuecomment-1486624308)

### CODE REVIEW

**Comment by jdksjolen** [comment]
Don't close yet.

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 1]
```diff
@@ -2171,7 +2171,7 @@ enum Nf {
     emit_int16(insn);                                                                        \
   }                                                                                          \
   void NAME(address dest) {                                                                  \
-    assert_cond(dest != NULL);                                                               \
+    assert_cond(dest != nullptr);                                                               \
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp:2174]: align

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 2]
```diff
@@ -2199,7 +2199,7 @@ enum Nf {
     emit_int16(insn);                                                                        \
   }                                                                                          \
   void NAME(Register Rs1, address dest) {                                                    \
-    assert_cond(dest != NULL);                                                               \
+    assert_cond(dest != nullptr);                                                               \
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp:2202]: align

File: src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp [hunk 3]
```diff
@@ -158,7 +158,7 @@ void ShenandoahBarrierSetAssembler::satb_write_barrier_pre(MacroAssembler* masm,
 
   // Calling the runtime using the regular call_VM_leaf mechanism generates
   // code (generated by InterpreterMacroAssember::call_VM_leaf_base)
-  // that checks that the *(rfp+frame::interpreter_frame_last_sp) == NULL.
+  // that checks that the *(rfp+frame::interpreter_frame_last_sp) == null.
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp:161]: is null

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp [hunk 4]
```diff
@@ -156,7 +156,7 @@ void InterpreterMacroAssembler::check_and_handle_earlyret(Register java_thread)
   if (JvmtiExport::can_force_early_return()) {
     Label L;
     ld(t0, Address(xthread, JavaThread::jvmti_thread_state_offset()));
-    beqz(t0, L);  // if [thread->jvmti_thread_state() == NULL] then exit
+    beqz(t0, L);  // if [thread->jvmti_thread_state() == null] then exit
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/interp_masm_riscv.cpp:159]: is

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp [hunk 5]
```diff
@@ -1319,19 +1319,19 @@ void InterpreterMacroAssembler::record_item_in_profile_helper(
...
-//     if (row[1].rec != NULL) then [
+//     if (row[1].rec != null) then [
 //       # degenerate decision tree, rooted at row[2]
 //       if (row[2].rec == rec) then [
 //         row[2].incr()
 //         goto done
 //       ]
-//       if (row[2].rec != NULL) then [
+//       if (row[2].rec != null) then [
 //         count.incr()
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/interp_masm_riscv.cpp:1335]: nullptr

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 6]
```diff
@@ -833,7 +833,7 @@ void MacroAssembler::li(Register Rd, int64_t imm) {
 
 #define INSN(NAME, REGISTER)                                       \
   void MacroAssembler::NAME(const address dest, Register temp) {   \
-    assert_cond(dest != NULL);                                     \
+    assert_cond(dest != nullptr);                                     \
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:836]: align

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 7]
```diff
@@ -877,7 +877,7 @@ void MacroAssembler::li(Register Rd, int64_t imm) {
 
 #define INSN(NAME)                                                                    \
   void MacroAssembler::NAME(Register Rd, const address dest, Register temp) {         \
-    assert_cond(dest != NULL);                                                        \
+    assert_cond(dest != nullptr);                                                        \
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:880]: align

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 8]
```diff
@@ -2022,13 +2022,13 @@ void MacroAssembler::access_load_at(BasicType type, DecoratorSet decorators,
...
 void MacroAssembler::null_check(Register reg, int offset) {
   if (needs_explicit_null_check(offset)) {
-    // provoke OS NULL exception if reg = NULL by
+    // provoke OS null exception if reg = null by
     // accessing M[reg] w/o changing any registers
     // NOTE: this is plenty to provoke a segv
     ld(zr, Address(reg, 0));
   } else {
     // nothing to do, (later) access of M[reg + offset]
-    // will provoke OS NULL exception if reg = NULL
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2031]: is

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 9]
```diff
@@ -2022,13 +2022,13 @@ void MacroAssembler::access_load_at(BasicType type, DecoratorSet decorators,
 
 void MacroAssembler::null_check(Register reg, int offset) {
   if (needs_explicit_null_check(offset)) {
-    // provoke OS NULL exception if reg = NULL by
+    // provoke OS null exception if reg = null by
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2025]: is

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 10]
```diff
@@ -2022,13 +2022,13 @@ void MacroAssembler::access_load_at(BasicType type, DecoratorSet decorators,
...
     // accessing M[reg] w/o changing any registers
     // NOTE: this is plenty to provoke a segv
     ld(zr, Address(reg, 0));
   } else {
     // nothing to do, (later) access of M[reg + offset]
-    // will provoke OS NULL exception if reg = NULL
+    // will provoke OS null exception if reg = null
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp:2031]: is

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 11]
```diff
@@ -590,7 +590,7 @@ class MacroAssembler: public Assembler {
 
 #define INSN(NAME)                                                                                       \
   void NAME(Register Rs1, Register Rs2, const address dest) {                                            \
-    assert_cond(dest != NULL);                                                                           \
+    assert_cond(dest != nullptr);                                                                           \
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:593]: align

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 12]
```diff
@@ -780,7 +780,7 @@ class MacroAssembler: public Assembler {
 
 #define INSN(NAME)                                                                                 \
   void NAME(Register Rd, address dest) {                                                           \
-    assert_cond(dest != NULL);                                                                     \
+    assert_cond(dest != nullptr);                                                                     \
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:783]: align

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 13]
```diff
@@ -837,7 +837,7 @@ class MacroAssembler: public Assembler {
 
 #define INSN(NAME)                                                                                 \
   void NAME(FloatRegister Rd, address dest, Register temp = t0) {                                  \
-    assert_cond(dest != NULL);                                                                     \
+    assert_cond(dest != nullptr);                                                                     \
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:840]: align

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 14]
```diff
@@ -897,7 +897,7 @@ class MacroAssembler: public Assembler {
 
 #define INSN(NAME)                                                                                 \
   void NAME(Register Rs, address dest, Register temp = t0) {                                       \
-    assert_cond(dest != NULL);                                                                     \
+    assert_cond(dest != nullptr);                                                                     \
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:900]: align

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 15]
```diff
@@ -943,7 +943,7 @@ class MacroAssembler: public Assembler {
 
 #define INSN(NAME)                                                                                 \
   void NAME(FloatRegister Rs, address dest, Register temp = t0) {                                  \
-    assert_cond(dest != NULL);                                                                     \
+    assert_cond(dest != nullptr);                                                                     \
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.hpp:946]: align

File: src/hotspot/cpu/riscv/relocInfo_riscv.cpp [hunk 16]
```diff
@@ -41,7 +41,7 @@ void Relocation::pd_set_data_value(address x, intptr_t o, bool verify_only) {
   switch (type()) {
     case relocInfo::oop_type: {
       oop_Relocation *reloc = (oop_Relocation *)this;
-      // in movoop when BarrierSet::barrier_set()->barrier_set_nmethod() != NULL
+      // in movoop when BarrierSet::barrier_set()->barrier_set_nmethod() != null
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/relocInfo_riscv.cpp:44]: isn't

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 17]
```diff
@@ -1765,21 +1765,21 @@ class StubGenerator: public StubCodeGenerator {
...
     // (4) length  must not be negative.
-    // (5) src klass and dst klass should be the same and not NULL.
+    // (5) src klass and dst klass should be the same and not null.
     // (6) src and dst should be arrays.
     // (7) src_pos + length must not exceed length of src.
     // (8) dst_pos + length must not exceed length of dst.
     //
 
-    // if [src == NULL] then return -1
+    // if [src == null] then return -1
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:1774]: is

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 18]
```diff
@@ -1765,21 +1765,21 @@ class StubGenerator: public StubCodeGenerator {
...
     __ beqz(src, L_failed);
 
     // if [src_pos < 0] then return -1
     // i.e. sign bit set
     __ andi(t0, src_pos, 1UL << 31);
     __ bnez(t0, L_failed);
 
-    // if [dst == NULL] then return -1
+    // if [dst == null] then return -1
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:1782]: is

File: src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp [hunk 19]
```diff
@@ -830,7 +830,7 @@ address TemplateInterpreterGenerator::generate_Reference_get_entry(void) {
 
   Label slow_path;
   const Register local_0 = c_rarg0;
-  // Check if local 0 != NULL
+  // Check if local 0 != null
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/templateInterpreterGenerator_riscv.cpp:833]: isn't

File: src/hotspot/cpu/riscv/templateTable_riscv.cpp [hunk 20]
```diff
@@ -3601,8 +3601,8 @@ void TemplateTable::instanceof() {
     __ bind(is_null);   // same as 'done'
   }
   __ bind(done);
-  // x10 = 0: obj == NULL or  obj is not an instanceof the specified klass
-  // x10 = 1: obj != NULL and obj is     an instanceof the specified klass
```
Comment (jdksjolen) [inline src/hotspot/cpu/riscv/templateTable_riscv.cpp:3605]: is isn't

**Review by jdksjolen — COMMENTED** [review]
Review of changes to be made.

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 21]
```diff
@@ -1833,7 +1833,7 @@ class StubGenerator: public StubCodeGenerator {
     __ load_klass(t1, dst);
     __ bne(t1, scratch_src_klass, L_failed);
 
-    // if [src->is_Array() != NULL] then return -1
+    // if [src->is_Array() != null] then return -1
```
Comment (dholmes-ora) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:1836]: nullptr for code fragment

**Review by dholmes-ora — APPROVED** [review]
LGTM!

One nit.

Thanks

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp [hunk 22]
```diff
@@ -1613,7 +1613,7 @@ void InterpreterMacroAssembler::call_VM_leaf_base(address entry_point,
    ld(t0, Address(fp, frame::interpreter_frame_last_sp_offset * wordSize));
    beqz(t0, L);
    stop("InterpreterMacroAssembler::call_VM_leaf_base:"
-        " last_sp != NULL");
+        " last_sp != null");
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/interp_masm_riscv.cpp:1616]: Maybe: " last_sp isn't null" ?

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp [hunk 23]
```diff
@@ -1640,7 +1640,7 @@ void InterpreterMacroAssembler::call_VM_base(Register oop_result,
     ld(t0, Address(fp, frame::interpreter_frame_last_sp_offset * wordSize));
     beqz(t0, L);
     stop("InterpreterMacroAssembler::call_VM_base:"
-         " last_sp != NULL");
+         " last_sp != null");
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/interp_masm_riscv.cpp:1643]: Similar here. Maybe: " last_sp isn't null" ?

**Review by RealFYang — APPROVED** [review]
LGTM.

File: src/hotspot/cpu/riscv/interpreterRT_riscv.cpp [hunk 24]
```diff
@@ -227,7 +227,7 @@ class SlowSignatureHandler
 
   virtual void pass_object() {
     intptr_t* addr = single_slot_addr();
-    intptr_t value = *addr == 0 ? NULL : (intptr_t)addr;
+    intptr_t value = *addr == 0 ? nullptr : (intptr_t)addr;
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/interpreterRT_riscv.cpp:230]: PS: I got compile errors when doing a native build with GCC-11.3.0:
```
474 /home/fyang/openjdk-jdk/src/hotspot/cpu/riscv/interpreterRT_riscv.cpp: In member function 'virtual void SlowSignatureHandler::pass_object()':
475 /home/fyang/openjdk-jdk/src/hotspot/cpu/riscv/interpreterRT_riscv.cpp:230:33: error: operands to '?:' have different types 'std::nullptr_t' and 'intptr_t' {aka 'long int'}
476   230 |     intptr_t value = *addr == 0 ? nullptr : (intptr_t)addr;
477       |                      ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
```

I think we should change the "nullptr" here into "(intptr_t)nullptr".

**Review by RealFYang — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp (hunk 22 above)
Comment (jdksjolen) [reply src/hotspot/cpu/riscv/interp_masm_riscv.cpp:1616]: Agreed.

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp (hunk 23 above)
Comment (jdksjolen) [reply src/hotspot/cpu/riscv/interp_masm_riscv.cpp:1643]: Agreed.

File: src/hotspot/cpu/riscv/interpreterRT_riscv.cpp (hunk 24 above)
Comment (jdksjolen) [reply src/hotspot/cpu/riscv/interpreterRT_riscv.cpp]: Yeah, that's correct. Fixed!

**Review by RealFYang — APPROVED** [review]
Update change looks good. Thanks.

**Comment by jdksjolen** [comment]
Thanks @RealFYang! Would you mind running the tier1 tests on RISC-V? I don't have access to that architecture.

**Comment by jdksjolen** [comment]
>org.opentest4j.AssertionFailedError: java.lang.ThreadLocal.set not found!!! ==> expected: <true> but was: <false>

linux-x86 test failure. Probably false alarm, as these changes shouldn't change the behaviour of linux-x86.

**Comment by RealFYang** [comment]
> Thanks @RealFYang! Would you mind running the tier1 tests on RISC-V? I don't have access to that architecture.

Sure! I have performed tier1 test on linux-riscv64 hifive unmatched boards, result looks good.

**Comment by jdksjolen** [comment]
Thanks!

/integrate

**Pull request closed without merging**